packetbeat/npcap/npcap.go

```diff
@@ -0,0 +1,284 @@
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+
+	"github.com/google/gopacket/pcap"
+	"golang.org/x/mod/semver"
+
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+// RegistryNotFound is returned by CurrentVersion if the network request
+// returns a 404 status code.
+var RegistryNotFound = errors.New("npcap: registry not found")
+
+var (
+	// Installer holds the embedded installer when run with x-pack.
+	Installer []byte
+
+	// EmbeddedInstallerVersion holds the version of the embedded installer.
+	EmbeddedInstallerVersion string
+)
+
+// Fetch downloads the Npcap installer, writes the content to the given filepath
+// and returns the sha256 hash of the downloaded object. If the registry is not
+// found RegistryNotFound is returned.
+func CurrentVersion(ctx context.Context, log *logp.Logger, registry string) (version, url, hash string, err error) {
+	if runtime.GOOS != "windows" {
+		return "", "", "", errors.New("npcap: called Fetch on non-Windows platform")
+	}
+	return currentVersion(ctx, log, registry)
+}
+
+func currentVersion(ctx context.Context, log *logp.Logger, registry string) (version, url, hash string, err error) {
+	req, err := http.NewRequestWithContext(ctx, "GET", registry, nil)
+	if err != nil {
+		return "", "", "", err
+	}
+
+	var client http.Client
+	res, err := client.Do(req)
+	if err != nil {
+		return "", "", "", err
+	}
+	defer res.Body.Close()
+
+	b, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		log.Errorf("failed to read the error response body: %v", err)
+	}
+	b = bytes.TrimSpace(b)
+	if len(b) == 0 {
+		// Give a meaningful error message and make json error if we have no body.
+		b = []byte("empty")
+	}
+	if res.StatusCode != http.StatusOK {
+		if res.StatusCode == http.StatusNotFound {
+			return "", "", "", fmt.Errorf("%w: %s", RegistryNotFound, registry)
+		}
+		if registry == "" {
+			registry = `""`
+		}
+		return "", "", "", fmt.Errorf("npcap: failed to fetch version info at %s, status: %d, message: %s", registry, res.StatusCode, b)
+	}
+
+	var info struct {
+		Version string
+		URL     string
+		Hash    string
+	}
+	err = json.Unmarshal(b, &info)
+	if err != nil {
+		return "", "", "", fmt.Errorf("%w: %#q", err, b)
+	}
+
+	return info.Version, info.URL, info.Hash, nil
+}
+
+// Fetch downloads the Npcap installer, writes the content to the given filepath
+// and returns the sha256 hash of the downloaded object.
+func Fetch(ctx context.Context, log *logp.Logger, url, path string) (hash string, err error) {
+	if runtime.GOOS != "windows" {
+		return "", errors.New("npcap: called Fetch on non-Windows platform")
+	}
+	return fetch(ctx, log, url, path)
+}
+
+func fetch(ctx context.Context, log *logp.Logger, url, path string) (hash string, err error) {
+	log.Infof("download %s to %s", url, path)
+
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	if err != nil {
+		return "", err
+	}
+
+	var client http.Client
```

Comment:
We should add User-Agent via libbeat/common/useragent.

Reply:
No longer relevant.